### ISSUE TITLE
Substring function returns zero precision

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#21]
```
What steps will reproduce the problem?

    // This is result for H2
    @Test
    public void testSubstringH2() throws Exception {

        Class.forName("org.h2.Driver");

        Connection ca = DriverManager.getConnection("jdbc:h2:mem:", "", "");

        Statement sa = ca.createStatement();

        sa.execute("CREATE TABLE one (C CHARACTER(12))");

        ResultSet rs = sa.executeQuery("SELECT substring(C,2,1) as D from
one;");
        ResultSetMetaData md = rs.getMetaData();
        assertEquals(md.getPrecision(1), 0);  // !!! return 0
        ca.close();
    }

    // This is result for DB2 for similar table
    @Test
    public void testSubstringDB2() throws Exception {

        Class.forName("COM.ibm.db2.jdbc.app.DB2Driver");

        Connection ca = DriverManager.getConnection("jdbc:db2:***", "***",
"***");

        Statement sa = ca.createStatement();

        ResultSet rs = sa.executeQuery("SELECT substr(C1,2,1) as D from
another.mytable;");
        ResultSetMetaData md = rs.getMetaData();
        assertEquals(md.getPrecision(1), 1); // !!! returns 1
        ca.close();
    }


What is the expected output? What do you see instead?

I would expect substring return 1 or the width of the first argument if the
third parameter is not constant.

What version of the product are you using? On what operating system, file
system, and virtual machine?

H2-02-02-08

Do you know a workaround?

No

How important/urgent is the problem for you?

Will try to fix it myself first

In your view, is this a defect or a feature request?

Defect

Please provide any additional information below.


```

Original issue reported on code.google.com by `pgane...@gmail.com` on 17 Apr 2008 at 1:05

**Comment by GoogleCodeExporter** [comment]
```
The fix is attached:

file Function.java:private void calculatePrecisionAndDisplaySize() {

add additional switch case

            case SUBSTR:
            case SUBSTRING:
                precision = 0;
                // if the third argument in constant use this value
                if (args.length == 3 && args[2] instanceof ValueExpression) {
                    Value a3 = ((ValueExpression) args[2]).getValue(null);
                    try {
                        precision = a3.getLong();
                    } catch (SQLException e) {
                        // TODO for TM:  throw proper exception here
                        throw new RuntimeException("substring 3 argument should be
integer");
                    }
                }
                // for two argument for constant offset substract offset from first
argument length
                else if (args.length == 2 && args[1] instanceof ValueExpression) {
                    Value a2 = ((ValueExpression) args[1]).getValue(null);
                    try {
                        precision = args[1].getPrecision() - a2.getLong() + 1;
                    } catch (SQLException e) {
                        // TODO for TM: throw proper exception here
                        throw new RuntimeException("substring 2 argument should be
integer");
                    }
                }
                // otherwise take conservative assumption that the lenght will be
equal to the first argument
                else {
                    precision = args[1].getPrecision();
                }
                displaySize = MathUtils.convertLongToInt(precision);
                break;

```

Original comment by `pgane...@gmail.com` on 17 Apr 2008 at 1:52

**Comment by GoogleCodeExporter** [comment]
```
Now DB2 test and H2 match: tests are attached:

    @Test
    public void testSubstringH2() throws Exception {

        Class.forName("org.h2.Driver");

        Connection ca = DriverManager.getConnection("jdbc:h2:mem:", "", "");

        Statement sa = ca.createStatement();

        sa.execute("CREATE TABLE mytable (C1 CHARACTER(10), I1 INTEGER)");


        {
            ResultSet rs = sa.executeQuery("SELECT C1 as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 10);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2,1) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 1);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 9);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2,I1) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 10);
        }
        ca.close();
    }

    @Test
    public void testSubstringDB2() throws Exception {

        Class.forName("COM.ibm.db2.jdbc.app.DB2Driver");

        Connection ca = DriverManager.getConnection("jdbc:db2:***, "***", "***");

        Statement sa = ca.createStatement();
        sa.execute("SET SCHEMA ANOTHER");

        {
            ResultSet rs = sa.executeQuery("SELECT C1 as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 10);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2,1) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 1);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 9);
        }
        {
            ResultSet rs = sa.executeQuery("SELECT substr(C1,2,I1) as D from mytable;");
            ResultSetMetaData md = rs.getMetaData();
            assertEquals(md.getPrecision(1), 10);
        }
        ca.close();
    }



```

Original comment by `pgane...@gmail.com` on 17 Apr 2008 at 1:54

**Comment by GoogleCodeExporter** [comment]
```
There was typo: resubmit the fix

            case SUBSTR:
            case SUBSTRING:
                precision = 0;
                // if the third argument in constant use this value
                if (args.length == 3 && args[2] instanceof ValueExpression) {
                    Value a3 = ((ValueExpression) args[2]).getValue(null);
                    try {
                        precision = a3.getLong();
                    } catch (SQLException e) {
                        // TODO for TM:  throw proper exception here
                        throw new RuntimeException("substring 3 argument should be
integer");
                    }
                }
                // for two argument for constant offset substract offset from first
argument length
                else if (args.length == 2 && args[1] instanceof ValueExpression) {
                    Value a2 = ((ValueExpression) args[1]).getValue(null);
                    try {
                        precision = args[0].getPrecision() - a2.getLong() + 1;
                    } catch (SQLException e) {
                        // TODO for TM: throw proper exception here
                        throw new RuntimeException("substring 2 argument should be
integer");
                    }
                }
                // otherwise take conservative assumption that the lenght will be
equal to the first argument
                else {
                    precision = args[0].getPrecision();
                }
                displaySize = MathUtils.convertLongToInt(precision);
                break;

```

Original comment by `pgane...@gmail.com` on 17 Apr 2008 at 2:04

**Comment by GoogleCodeExporter** [comment]
```
My plan is to implement it like this:
        case SUBSTRING:
        case SUBSTR: {
            precision = args[0].getPrecision();
            if (args[1].isConstant()) {
                // if only two arguments are used, 
                // subtract offset from first argument length
                precision -= args[1].getValue(session).getLong() - 1;
            }
            if (args.length == 3 && args[2].isConstant()) {
                // if the third argument is constant it is at most this value
                precision = Math.min(precision, args[2].getValue(session).getLong());
            }
            precision = Math.max(0, precision);
            displaySize = MathUtils.convertLongToInt(precision);
            break;
        }

My test is:

    private void testSubstringPrecision() throws Exception {
        trace("testSubstringPrecision");
        stat.execute("CREATE TABLE TEST(ID INT, NAME VARCHAR(10))");
        stat.execute("INSERT INTO TEST VALUES(1, 'Hello'), (2, 'WorldPeace')");
        checkPrecision(9, "SELECT SUBSTR(NAME, 2) FROM TEST");
        checkPrecision(10, "SELECT SUBSTR(NAME, ID) FROM TEST");
        checkPrecision(4, "SELECT SUBSTR(NAME, 2, 4) FROM TEST");
        checkPrecision(0, "SELECT SUBSTR(NAME, 12, 4) FROM TEST");
        checkPrecision(3, "SELECT SUBSTR(NAME, 8, 4) FROM TEST");
        checkPrecision(4, "SELECT SUBSTR(NAME, 7, 4) FROM TEST");
        checkPrecision(8, "SELECT SUBSTR(NAME, 3, ID*0) FROM TEST");
        stat.execute("DROP TABLE TEST");
    }

    private void checkPrecision(int expected, String sql) throws Exception {
        ResultSetMetaData meta = stat.executeQuery(sql).getMetaData();
        check(expected, meta.getPrecision(1));
    }

PostgreSQL and HSQLDB doesn't support it. 
Both MySQL and Derby support it up to some point (not always correctly).

By the way, why do you need this feature?
```

Original comment by `thomas.t...@gmail.com` on 17 Apr 2008 at 5:11
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
Original comment by `thomas.t...@gmail.com` on 17 Apr 2008 at 5:11

**Comment by GoogleCodeExporter** [comment]
```
Now H2 matches Oracle in this aspect. DB2 gives errors and/or different results

    @Test
    public void testSubstringForDB2() throws Exception {

        Class.forName("COM.ibm.db2.jdbc.app.DB2Driver");
        Connection ca = DriverManager.getConnection("jdbc:db2:sample", "**", "**");

        Statement stat = ca.createStatement();

        try {
            stat.execute("DROP TABLE TEST");
        } catch (SQLException e) {
        }

        stat.execute("CREATE TABLE TEST(ID INT, NAME VARCHAR(10))");
        stat.execute("INSERT INTO TEST VALUES(1, 'Hello'), (2, 'WorldPeace')");

        checkPrecision(10, "SELECT SUBSTR(NAME, 2) FROM TEST", stat);  // DIFFERENT
        checkPrecision(10, "SELECT SUBSTR(NAME, ID) FROM TEST", stat);
        checkPrecision(4, "SELECT SUBSTR(NAME, 2, 4) FROM TEST", stat);

        try {
            checkPrecision(0, "SELECT SUBSTR(NAME, 12, 4) FROM TEST", stat); //
runtime error
            fail("error expected");
        } catch (Exception e) {
        }

        try {
            checkPrecision(3, "SELECT SUBSTR(NAME, 8, 4) FROM TEST", stat);  //
runtime error
            fail("error expected");
        } catch (Exception e) {
        }

        checkPrecision(4, "SELECT SUBSTR(NAME, 7, 4) FROM TEST", stat);
        checkPrecision(10, "SELECT SUBSTR(NAME, 3, ID*0) FROM TEST", stat); // DIFFERENT

        stat.execute("DROP TABLE TEST");
        ca.close();

    }

    @Test
    public void testSubstringOracle() throws Exception {

        String driver = "oracle.jdbc.driver.OracleDriver";
        Class.forName(driver);
        String url = "jdbc:oracle:thin:@***:1521:XE";
        Connection ca = DriverManager.getConnection(url, "scott", "tiger");

        Statement stat = ca.createStatement();

        try {
            stat.execute("DROP TABLE TEST");
        } catch (SQLException e) {
        }

        stat.execute("CREATE TABLE TEST(ID INT, NAME VARCHAR(10))");
        stat.execute("INSERT INTO TEST VALUES(1, 'Hello')");
        stat.execute("INSERT INTO TEST VALUES (2, 'WorldPeace')");

        checkPrecision(9, "SELECT SUBSTR(NAME, 2) FROM TEST", stat);
        checkPrecision(10, "SELECT SUBSTR(NAME, ID) FROM TEST", stat);
        checkPrecision(4, "SELECT SUBSTR(NAME, 2, 4) FROM TEST", stat);
        checkPrecision(0, "SELECT SUBSTR(NAME, 12, 4) FROM TEST", stat);
        checkPrecision(3, "SELECT SUBSTR(NAME, 8, 4) FROM TEST", stat);  
        checkPrecision(4, "SELECT SUBSTR(NAME, 7, 4) FROM TEST", stat);
        checkPrecision(8, "SELECT SUBSTR(NAME, 3, ID*0) FROM TEST", stat);

        stat.execute("DROP TABLE TEST");
        ca.close();

    }

```

Original comment by `pgane...@gmail.com` on 17 Apr 2008 at 7:59

**Comment by GoogleCodeExporter** [comment]
```
Fixed in version 1.0.70 (2008-04-20)
```

Original comment by `thomas.t...@gmail.com` on 21 Apr 2008 at 5:48
- Changed state: **Fixed**
